tools/dev.js

```diff
@@ -0,0 +1,87 @@
+const path = require("path");
+const { createServer } = require("http");
+const chokidar = require("chokidar");
+const karma = require("karma");
+const serve = require("serve-handler");
+const colors = require("colors");
+const serveConfig = require("../serve.json");
+const { Builder } = require("./builder");
+
+const KARMA_PORT = 9876;
+const SERVE_PORT = 5000;
+
+const karmaConfig = karma.config.parseConfig(
+  path.join(__dirname, "../karma.conf.js"),
+  {
+    browsers: ["ChromeHeadless"],
+    autoWatch: false,
+    port: KARMA_PORT,
+    logLevel: karma.constants.LOG_WARN,
+  }
+);
+const karmaServer = new karma.Server(karmaConfig);
+
+main();
+
+function main() {
+  const server = createServer((req, res) => serve(req, res, serveConfig));
+  server.listen(SERVE_PORT);
+  server.on("error", onError);
+
+  karmaServer.start();
+
+  const watcher = chokidar.watch("./src", { ignoreInitial: true });
+  watcher.on("all", onFileChange);
+  watcher.on("error", onError);
+
+  printWelcomeMessage();
+}
+
+async function onFileChange(_event, _file) {
+  try {
+    await Builder.build({ name: "w3c", debug: true });
+    karma.runner.run({ port: KARMA_PORT, autoWatch: true }, () => {});
```

Comment:
Can we cancel the previous run when file changes happen but the tests are not finished yet?

Reply:
Couldn't figure that one out. Will try again

Reply:
So the karma API allows to stop the server, but doesn't allow stopping a run in between. One way could be to stop the server so the run fails and restart the server again.

Reply:
That sounds reasonable enough.

Reply:
Didn't work :smile: 
Want to give it a try? https://karma-runner.github.io/4.0/dev/public-api.html

Reply:
karmaServer is in EventEmitter :tada: 
``` js
karmaServer.emit("run_complete", [], {});
```